feat(hono/context): add buffer returns

askorupskyy opened the pull request:
# Context

I needed to be able to return Buffers from my Hono app, and was not able to do so.

Resolves #3729 #3720

### The author should do the following, if applicable

- [x] Add tests
- [x] Run tests
- [x] `bun run format:fix && bun run lint:fix` to format the code
- [x] Add [TSDoc](https://tsdoc.org/)/[JSDoc](https://jsdoc.app/about-getting-started) to document the code

askorupskyy commented:
Tested this PR with the sample repo, using `typescript@5.7` – seems to be working fine

<img width="1465" alt="Screenshot 2025-01-08 at 6 18 47 PM" src="https://github.com/user-attachments/assets/ed375d94-a7e9-418d-9983-811d3c8c0e88" />

yusukebe commented on src/context.ts:
```diff
@@ -1,3 +1,5 @@
+import type { Buffer } from 'node:buffer'
```
Is this `import type` necessary? In my environment, it does not throw an error without this.

askorupskyy replied:
I added this for Deno. Deno does not have a Buffer exposed to the global scope

yusukebe replied:
It is importing just types, but I think we should not use `node:*` modules in the core of `hono`. Actually, `Buffer` is not in Web Standards API.

askorupskyy replied:
I think using `node:buffer` is fine here, especially as a Type.  There's no way to polyfill types for specific platforms as far as I am aware (so Deno will continue to fail). In addition to that, `node:buffer` is either the default implementation for a lot of runtimes, such as Cloudflare Workers, Deno, etc, or is already defined as a global replicating Node.js API. I see no problem with using the type from Node, since all runtimes are pretty much using the same API as node.

EdamAme-x replied:
It might be a good option to have a helper function to convert from `node:buffer` to the web standard accepted body format (For example, `ReadbleStream`).
If you think it is a good idea, we can change the implementation policy of this PR.

askorupskyy replied:
Let me see if i can make this work with a uint8array

askorupskyy commented:
@yusukebe @EdamAme-x looks like `Buffer` indeed is a Node-only construct (they made it before Uint8Array was invented), so to be compliant with Web Standards I moved to `Uint8Array`, for which the Buffer is a superset. Added one more test and it looks like everything works

EdamAme-x reviewed (approved):
a

EdamAme-x commented:
One thing I was wondering, is there a reason why `URLSearchParams` and `FormData` are not accepted?

[TypeScript Playground](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgCQgOwnAvnAZlCEOAIgAt0JiAoKgY3QGd4BDMMOAXjjQFMB3FBQAUAShqswAOgDmPGENqcAfIipx1cKHICuUNHFqSARhAAmATyG8BAVQBKAGQDKPZlFqkACm+YgGQgG01DRCA4gZtWloeBgZiABo4GChtHgBdYPURMSwRIA)

askorupskyy commented:
@EdamAme-x first of all, `BodyInit` is what's used by `ClientRequestImpl`, which is part `hc`. That's to parse whatever input you give to `hc` – this is why your second example doesn't work. I don't think BodyInit matters in the context of this PR

askorupskyy commented:
@yusukebe can you review this once again pls?

yusukebe reviewed (approved):
LGTM!

yusukebe commented:
@askorupskyy 

Looks good to me! We can merge this. I'm considering whether this change should be included in a patch release or a minor release. I'll decide later. Thanks!
